File: src/main/python/com/hazza/leetcode_by_python/SegmentOne/No35_Search_Insert_Position.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/28 20:10
# https://leetcode.com/problems/search-insert-position/
from typing import List


class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        if target <= nums[l]:
            return 0
        if target > nums[r]:
            return r + 1
        while l <= r:
            mid = l + (r - l) // 2
            if nums[mid] == target:
                return mid
            if nums[mid] < target:
                l = mid + 1
            else:
                r = mid - 1

        return l


if __name__ == '__main__':
    solution = Solution()

    assert solution.searchInsert([1, 3, 5, 6], 5) == 2
    assert solution.searchInsert([1, 3, 5, 6], 2) == 1
    assert solution.searchInsert([1, 3, 5, 6], 7) == 4
    assert solution.searchInsert([1, 3, 5, 6], 0) == 0

File: src/main/java/com/hazza/leetcode/SegmentOne/No35SearchInsertPosition.java
package com.hazza.leetcode.SegmentOne;

/**
 * https://leetcode.com/problems/search-insert-position/#/description
 */
public class No35SearchInsertPosition {
    public int searchInsert(int[] nums, int target) {
        if (nums == null || nums.length == 0) return -1;
        int left = 0, right = nums.length - 1;
        if (target < nums[left]) return left;
        if (target > nums[right]) return right + 1;
        while (left <= right) {
            int middle = (left + right) / 2;
            if (nums[middle] == target) return middle;
            if (nums[middle] > target) right = middle - 1;
            else left= middle + 1;
        }
        return left;
    }
}
